### PR TITLE
Total order of binding times across all units in Aliases

### PR DESCRIPTION
**Pull request by mshinwell**
As discussed on Slack.  This should remove the prohibition on adding aliases where both sides are imported variables.  Hopefully this will fix the bug @lukemaurer has been looking at.

### CODE REVIEW

File: middle_end/flambda2/types/env/typing_env.ml [hunk 1]
```diff
@@ -211,6 +212,21 @@ module Join_env = struct
     Meet_env.already_meeting central_env simple1 simple2
 end
 
+let names_to_types t =
+  Cached_level.names_to_types (One_level.just_after_level t.current_level)
+
+exception Binding_time_resolver_failure
+
+let binding_time_resolver resolver name =
+  match resolver (Name.compilation_unit name) with
```
Comment (lthls) [inline middle_end/flambda2/types/env/typing_env.ml:221]: I think you should catch exceptions coming from the call to `resolver` and translate them to a fatal error, similar to what `find_with_binding_time_and_mode'` is doing (it helps with debugging)

**Review by lthls — APPROVED** [review]
This looks fine, but I also thought of an alternative simpler solution while reviewing this (see the discussion on Slack). I've also left two comments for possible improvements.

File: middle_end/flambda2/types/env/aliases.ml [hunk 2]
```diff
@@ -407,20 +408,42 @@ let [@ocamlformat "disable"] print ppf { canonical_elements; aliases_of_canonica
...
+    let than_comp_unit = Name.compilation_unit than in
+    (* The compilation unit ordering is arbitrary, but total. *)
+    let c = Compilation_unit.compare alias_comp_unit than_comp_unit in
+    if c < 0
+    then true
+    else if c > 0
+    then false
+    else if Compilation_unit.equal alias_comp_unit
+              (Compilation_unit.get_current_exn ())
+    then false
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/types/env/aliases.ml:426]: Fixed

File: middle_end/flambda2/types/env/typing_env.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/flambda2/types/env/typing_env.ml:221]: Done

**Pull request merged**